### PR TITLE
Refactor `CommonCodeSystemTerminologyService` method `validateCode`

### PR DESCRIPTION
**Pull request by codeforgreen**
Found that `CommonCodeSystemTerminologyService` method `validateCode` was hard to follow and inconsistent across different types of supported `CodeSystem`. 

What was done:
- Create common methods for creating validation output object, for success and failure
- Create common method for retrieving localized error message
- Extract method for mapping `ValueSet` url to the corresponding `CodeSystem` url
- Check that `CodeSystem` matches uniformly across all types
- Use infer system option uniformly across all types
- Add a bunch of new test cases

Bonus:
- Renamed some test classes to reflect the tested service. Npm vs PackageInstallerSvc.

### CODE REVIEW

File: hapi-fhir-validation/src/main/java/org/hl7/fhir/common/hapi/validation/support/CommonCodeSystemsTerminologyService.java [hunk 1]
```diff
@@ -108,136 +108,108 @@ public CodeValidationResult validateCodeInValueSet(
...
-							.setDisplay(match.get().getDisplay());
+		switch (valueSet) {
+			case USPS_VALUESET_URL:
+				if (USPS_CODES.containsKey(theCode)) {
+					return getValidateCodeResultOk(theCode, USPS_CODES.get(theCode));
 				} else {
-					return new CodeValidationResult()
-							.setSeverity(IssueSeverity.ERROR)
-							.setMessage("Code \"" + theCode + "\" is not in valueset: " + theValueSetUrl);
+					return getValidateCodeResultInError("unknownCodeInSystem", system, theCode);
```
Comment (volodymyr-korzh) [inline hapi-fhir-validation/src/main/java/org/hl7/fhir/common/hapi/validation/support/CommonCodeSystemsTerminologyService.java:136]: Suggestion: You can refactor this further - by extracting logic for USPS_VALUESET_URL and CURRENCIES_VALUESET_URL cases:
```java
case USPS_VALUESET_URL:
	return getCodeValidationResult(USPS_CODES, theCode, system);
case CURRENCIES_VALUESET_URL:
	return getCodeValidationResult(ISO_4217_CODES, theCode, system);
```

ExtractedMethod:
```java
private CodeValidationResult getCodeValidationResult(Map<String, String> theCodesMap, String theCode, String theSystem) {
		if (theCodesMap.containsKey(theCode)) {
			return getValidateCodeResultOk(theCode, theCodesMap.get(theCode));
		} else {
			return getValidateCodeResultInError("unknownCodeInSystem", theSystem, theCode);
		}
	}
```

File: hapi-fhir-validation/src/main/java/org/hl7/fhir/common/hapi/validation/support/CommonCodeSystemsTerminologyService.java [hunk 2]
```diff
@@ -108,136 +108,108 @@ public CodeValidationResult validateCodeInValueSet(
...
+			case UCUM_VALUESET_URL:
+				theCodeSystem = UCUM_CODESYSTEM_URL;
+				break;
 		}
+		return theCodeSystem;
+	}
 
-		return null;
+	protected CodeValidationResult getValidateCodeResultInError(
+			final String errorCode, final String param0, final String param1) {
```
Comment (volodymyr-korzh) [inline hapi-fhir-validation/src/main/java/org/hl7/fhir/common/hapi/validation/support/CommonCodeSystemsTerminologyService.java:187]: Nitpick: Please rename parameter names (errorCode, param0, param1):

`getValidateCodeResultInError(final String theErrorCode, final String theFirstParam, final String theSecondParam)`

File: hapi-fhir-validation/src/main/java/org/hl7/fhir/common/hapi/validation/support/CommonCodeSystemsTerminologyService.java [hunk 3]
```diff
@@ -1373,4 +1352,8 @@ private static HashMap<String, String> buildIso3166Codes() {
 		codes.put("ZWE", "Zimbabwe");
 		return codes;
 	}
+
+	protected String getErrorMessage(String errorCode, String param0, String param1) {
```
Comment (volodymyr-korzh) [inline hapi-fhir-validation/src/main/java/org/hl7/fhir/common/hapi/validation/support/CommonCodeSystemsTerminologyService.java:1356]: NitPick: rename parameter
`(final String theErrorCode, final String theFirstParam, final String theSecondParam)`

File: hapi-fhir-validation/src/main/java/org/hl7/fhir/common/hapi/validation/support/CommonCodeSystemsTerminologyService.java [hunk 4]
```diff
@@ -108,136 +108,108 @@ public CodeValidationResult validateCodeInValueSet(
...
+				theCodeSystem = UCUM_CODESYSTEM_URL;
+				break;
 		}
+		return theCodeSystem;
+	}
 
-		return null;
+	protected CodeValidationResult getValidateCodeResultInError(
+			final String errorCode, final String param0, final String param1) {
+		String message = myFhirContext.getLocalizer().getMessage(getClass(), errorCode, param0, param1);
```
Comment (volodymyr-korzh) [inline hapi-fhir-validation/src/main/java/org/hl7/fhir/common/hapi/validation/support/CommonCodeSystemsTerminologyService.java:188]: Nitpick: you can reuse `getErrorMessage` method

`String message = getErrorMessage(theErrorCode, theFirstParam, theSecondParam);`

File: hapi-fhir-validation/src/main/java/org/hl7/fhir/common/hapi/validation/support/CommonCodeSystemsTerminologyService.java [hunk 5]
```diff
@@ -108,136 +108,108 @@ public CodeValidationResult validateCodeInValueSet(
...
+		return getValidateCodeResultError(message);
+	}
+
+	protected CodeValidationResult getValidateCodeResultOk(final String theCode, final String theDisplay) {
+		return new CodeValidationResult().setCode(theCode).setDisplay(theDisplay);
+	}
+
+	protected CodeValidationResult getValidateCodeResultError(final String message) {
```
Comment (volodymyr-korzh) [inline hapi-fhir-validation/src/main/java/org/hl7/fhir/common/hapi/validation/support/CommonCodeSystemsTerminologyService.java:196]: Nitpick: (`final String theMessage)`

File: hapi-fhir-validation/src/main/java/org/hl7/fhir/common/hapi/validation/support/CommonCodeSystemsTerminologyService.java [hunk 6]
```diff
@@ -81,10 +82,9 @@ public class CommonCodeSystemsTerminologyService implements IValidationSupport {
 	 * Constructor
 	 */
 	public CommonCodeSystemsTerminologyService(FhirContext theFhirContext) {
-		Validate.notNull(theFhirContext);
-
 		myFhirContext = theFhirContext;
 		myVersionCanonicalizer = new VersionCanonicalizer(theFhirContext);
+		Objects.requireNonNull(theFhirContext);
```
Comment (volodymyr-korzh) [inline hapi-fhir-validation/src/main/java/org/hl7/fhir/common/hapi/validation/support/CommonCodeSystemsTerminologyService.java:87]: Question: should we move it up as a first line of the constructor ?

File: hapi-fhir-validation/src/test/java/org/hl7/fhir/common/hapi/validation/support/CommonCodeSystemsTerminologyServiceTest.java [hunk 7]
```diff
@@ -31,214 +49,448 @@ public void before() {
...
-		assert outcome != null;
+	@ParameterizedTest
+	@CsvSource({"en-CA, English Canada", "en-US, English United States"})
+	public void testValidateLookupCode_withLanguagesWithKnownCode_returnsValid(final String theCode, final String theDisplay) {
+		CodeValidationResult outcome = mySvc.validateLookupCode(newSupport(), theCode, LANGUAGES_CODESYSTEM_URL);
+		assertNotNull(outcome);
 		assertTrue(outcome.isOk());
-		assertEquals("English Canada", outcome.getDisplay());
+		assertEquals(theCode, outcome.getCode());
+		assertEquals(theDisplay, outcome.getDisplay());
```
Comment (volodymyr-korzh) [inline hapi-fhir-validation/src/test/java/org/hl7/fhir/common/hapi/validation/support/CommonCodeSystemsTerminologyServiceTest.java:138]: Suggestion:
You can extract these validations to separate method, something like `validateOutcomeIsOk`.

These validations repeats 7 times in the test.
```java
		assertNotNull(outcome);
		assertTrue(outcome.isOk());
		assertEquals(theCode, outcome.getCode());
		assertEquals(theDisplay, outcome.getDisplay());
```

File: hapi-fhir-validation/src/test/java/org/hl7/fhir/common/hapi/validation/support/CommonCodeSystemsTerminologyServiceTest.java [hunk 8]
```diff
@@ -31,214 +49,448 @@ public void before() {
...
 
+	@Test
+	public void testFetchCodeSystem_withMimeType_returnsOk() {
+		CodeSystem cs = (CodeSystem) mySvc.fetchCodeSystem(MIMETYPES_CODESYSTEM_URL);
+		assertNull(cs);
+	}
+
+	@ParameterizedTest
+	@ValueSource(strings = { EncodingEnum.JSON_PLAIN_STRING, Constants.CT_FHIR_JSON_NEW, Constants.CT_FHIR_JSON })
+	public void testValidateCode_withMimetypesValueSetWithStandardCode_returnsValid(String code) {
```
Comment (volodymyr-korzh) [inline hapi-fhir-validation/src/test/java/org/hl7/fhir/common/hapi/validation/support/CommonCodeSystemsTerminologyServiceTest.java:376]: NitPick: (same for tests below)
`(String code) -> (String theCode) `

**Review by volodymyr-korzh — APPROVED** [review]
Approved with minor suggestions. 
Looks Good. Great work on refactoring!

File: hapi-fhir-validation/src/main/java/org/hl7/fhir/common/hapi/validation/support/CommonCodeSystemsTerminologyService.java (hunk 6 above)
Comment (codeforgreen) [reply hapi-fhir-validation/src/main/java/org/hl7/fhir/common/hapi/validation/support/CommonCodeSystemsTerminologyService.java:87]: Makes sense. I think I did but I got a compile warning. Let me double check this.

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/packages/PackageInstallerSvcR4Test.java [hunk 9]
```diff
@@ -162,6 +158,21 @@ public void testInstallUsCore() {
 		myPackageInstallerSvc.install(spec);
 	}
 
+	@Disabled("This test was added to validate IG installation scenario for OH")
```
Comment (codeforgreen) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/packages/PackageInstallerSvcR4Test.java:161]: Sorry, I will remove this test when I merge, it is not needed in this MR.

File: hapi-fhir-validation/src/main/java/org/hl7/fhir/common/hapi/validation/support/CommonCodeSystemsTerminologyService.java (hunk 6 above)
Comment (codeforgreen) [reply hapi-fhir-validation/src/main/java/org/hl7/fhir/common/hapi/validation/support/CommonCodeSystemsTerminologyService.java:87]: Yeah that works.
